Whatwg url object fix from will (informal review)

notmatt opened the pull request.

Commit:
rework of https.request, allows for WHATWG-URL obj

The honeycomb https.request wrapper instrumentation can't handle some of
the WHATWG-URL cases. For example, if you pass in a URL obj, options
(another object), and a callback, it will assume the first arg is the
options and the second is the callback. This will overwrite some of the
values (like the domain) and the call will fail.

For more info, see: https://nodejs.org/api/https.html#https_https_request_url_options_callback

Commit:
Building the url to send in the HC request

Sometimes we have a url as a string, sometimes we have a URL object, and
sometimes we just have a plain options object. Build the url for the
Honeycomb span based on which we get.

Commit:
Building the url to send in the HC request

Sometimes we have a url as a string, sometimes we have a URL object, and
sometimes we just have a plain options object. Build the url for the
Honeycomb span based on which we get.

Commit:
applying whatwg url fixes to http.js as well

Commit:
Merge branch 'whatwgUrlObjectFix' of github.com:wcgallego/beeline-nodejs into whatwgUrlObjectFix

Commit:
fixing lint

Left in a tracker (whoops) and missing semicolons

Commit:
fixing https tests

Partially clean up of the tests themselves. As for the https.js and
http.js instrumentation, checking for port and path before assuming
their values as it was giving us undefined if, for example, the path
wasn't set.

notmatt commented on lib/instrumentation/http.js:
```diff
@@ -20,47 +20,85 @@ let instrumentHTTP = (http, opts = {}) => {
+    // http.request(url[, options][, callback])
+    // So we can have 1-3 args to work with. For more details, see:
+    // http://nodejs.org/api/http.html#http_http_request_url_options_callback
+    return function (/* Args are parsed using the local built-in variable "arguments" */) {
```
Looks like HC targets node 10+, we could use [Rest Parameters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters#browser_compatibility) here. Not sure if they have a preference around that (I do not), but `arguments` is arguably a bit of a wart in JS, because it's an "array-like object" and that can bite you or add boilerplate like the `Array.from(arguments)` when you need to to actually be an array.

wcgallego replied:
yeah. this isn't a bad idea! I appreciate the subtlety with the built in vs. a variadic set of args.

notmatt replied:
Oh yes, you can still have named parameters when using `arguments.length` to check arity. I might lean that way to match the naming in the node docs.

I'd also name this anon function (`wrappedRequest` or similar) because it makes debugging easier, but they are clearly not fussed about that.

Commit:
using restParams instead

variadic args makes it a little cleaner to work with the args, as it's
then an actual array we're working with instead of the arguments built
in which is array *like*

notmatt commented on lib/instrumentation/http.js:
```diff
@@ -20,47 +20,85 @@ let instrumentHTTP = (http, opts = {}) => {
+    return function (..._args) {
       if (!api.traceActive()) {
-        return original.apply(this, Array.from(arguments));
+        return original.apply(this, Array.from(_args));
```
We can drop the `Array.from` now because variadic args are proper arrays already.
```javascript
function arrrrr(...args) {
    console.log("arguments is an array? ", Array.isArray(arguments))
    console.log("_args is an array? ", Array.isArray(_args))
}
arrrrr()
// arguments is an array?  false
// _args is an array?  true
```

notmatt commented on lib/instrumentation/http.js:
```diff
@@ -20,47 +20,85 @@ let instrumentHTTP = (http, opts = {}) => {
+        // If we have more, someone's doing something *weird* but account for it anyway
+        _url = _args[0];
+        combinedOptions = _args[1];
+        callback = _args[2];
```
If you wanted to be fancy, could use array destructuring here; `[_url, combinedOptions, callback] = _args` but I don't know how fancy you like your JS. 💰 🍾 🍸

notmatt commented on lib/instrumentation/http.js:
```diff
@@ -20,47 +20,85 @@ let instrumentHTTP = (http, opts = {}) => {
+        // _url, options
+
+        // First, determine if we have a callback or not, which can only be the second arg
+        if (typeof _args[1] === "function") {
```
➕ this is definitely the check they're missing in theirs. And I still think explicitly checking arity, then type is the better way of handling this weird situation.

notmatt commented on lib/instrumentation/http.js:
```diff
@@ -20,47 +20,85 @@ let instrumentHTTP = (http, opts = {}) => {
-          callback = options;
+          // Since we don't have a callback, the first arg must be a url, the second options
+          _url = _args[0];
+          combinedOptions = _args[1];
```
Yay, the part that was messing with us!

notmatt commented on lib/instrumentation/http.js:
```diff
@@ -20,47 +20,85 @@ let instrumentHTTP = (http, opts = {}) => {
+          _url = _args[0];
+          combinedOptions = _args[1];
+        }
+      } else if (_args.length === 1) {
```
I like that we have this explicitly; it was perfectly fine in the previous code, just implicit in whether or not `callback` was ever defined, and ambiguously "assigning" something we know is sometimes-but-not-always undefined, and then only checking that many lines later gives me the creeping horrors.

notmatt commented on lib/instrumentation/http.js:
```diff
@@ -20,47 +20,85 @@ let instrumentHTTP = (http, opts = {}) => {
+      if (_url) {
+        asyncSpanURL = _url;
+      } else if (combinedOptions instanceof URL) {
+        asyncSpanURL = url.format(new URL(combinedOptions), { auth: false }); // auth:false makes sure we don't leak user/pass
```
Good catch on dropping basic auth from going to HC! 🎖️

notmatt commented on lib/instrumentation/http.test.js:
```diff
@@ -51,7 +51,7 @@ test("url as a string", (done) => {
         [schema.EVENT_TYPE]: "http",
         [schema.TRACE_SPAN_NAME]: "GET",
-        url: "http://localhost:9009/",
+        url: "http://localhost:9009",
```
I'd mention why you're dropping those terminal `/`s.

notmatt commented on lib/instrumentation/http.test.js:
```diff
@@ -142,7 +142,7 @@ test("node 10.9.0+ url + options", (done) => {
 
   http.get(
-    "http://localhost:80",
+    "http://localhost:9009",
```
I _think_ they're trying to test that the port in `options` args will replace the port in the `url` here; might want to change it back? See [their lines 37-38ish](https://github.com/notmatt/beeline-nodejs/pull/1/files#diff-a04e82062dd81ac97fa3e7865c8b163aec88621d85032ceadcc142841db4c02aL37-L38).

notmatt commented on lib/instrumentation/https.test.js:
```diff
@@ -157,7 +157,7 @@ test("node 10.9.0+ url + options", (done) => {
 
   https.get(
-    "https://google.com:80",
+    "https://google.com:443",
```
Same hunch as the `http` tests; I think they're testing priority on what sets the port.

notmatt commented on lib/instrumentation/https.js:
```diff
@@ -84,18 +122,27 @@ let instrumentHTTPS = (https, opts = {}) => {
           } else {
             // the user didn't specify a callback, add it as a "response" handler ourselves
-            return original.apply(this, [combinedOptions]).on("response", wrapped_cb);
+            return original.apply(this, returnArgs).on("response", wrapped_cb);
```
Parallel implementation to the `http` version 👍🏻

notmatt reviewed (commented):
I'd say get rid of `Array.from()` and it's good to go!

From my digging through, the problem areas for them are:
L42: `options` is not guaranteed to be a function here.
L47: ditto

Which means that L75 we can end up calling `apply` on not-a-function (which is exactly what we saw, so hooray, fits with your analysis & the evidence).

And nice job on dropping basic auth from getting sent to HC.

Commit:
left in an Array.from, removing

vreynolds committed:
Add test for http request args
